proposals/0204-slashable-event-verification.md

```diff
@@ -0,0 +1,297 @@
+In order to verify that `shred1` and `shred2` were correctly signed by
+`node_pubkey` we use instruction retrospection.
+
+Using the `Instructions` sysvar we verify that the previous two instructions of
+this transaction are for the program ID
+`Ed25519SigVerify111111111111111111111111111`
+
+For each of these instructions, verify the instruction data:
+
+- The first byte is `0x01`
+- The second byte (padding) is `0x00`
+
+And then deserialize the remaining instruction data as 2 byte little-endian
+unsigned integers:
+
+```rust
+struct Ed25519SignatureOffsets {
+    signature_offset: u16,             // offset to ed25519 signature of 64 bytes
+    signature_instruction_index: u16,  // instruction index to find signature
+    public_key_offset: u16,            // offset to public key of 32 bytes
+    public_key_instruction_index: u16, // instruction index to find public key
+    message_data_offset: u16,          // offset to start of message data
+    message_data_size: u16,            // size of message data
+    message_instruction_index: u16,    // index of instruction data to get message
+                                       // data
+}
+```
+
+We wish to verify that these instructions correspond to
+
+```
+verify(pubkey = node_pubkey, message = shred1.merkle_root, signature = shred1.signature)
+verify(pubkey = node_pubkey, message = shred2.merkle_root, signature = shred2.signature)
+```
+
+We use the deserialized offsets to calculate [\[3\]](#notes) the `pubkey`,
+`message`, and `signature` of each instruction and verify that they correspond
+to the `node_pubkey`, `merkle_root`, and `signature` specified by the shred payload.
+
+If both proof and signer verification succeed, we continue on to store the incident.
+
+---
+
+### Incident reporting
+
+After verifying a successful proof we store the results in a program derived
+address for future use. The PDA is derived using the `node_pubkey`, `slot`, and
+the violation type:
+
+```rust
+let (pda, _) = find_program_address(&[
+  node_pubkey.to_bytes(),
+  slot.to_le_bytes(),
+  ViolationType::DuplicateBlock.to_u8(),
+])
+```
+
+At the moment `DuplicateBlock` is the only violation type but future work will
+add additional slashing types.
+
+If the `pda` account has non-zero lamports, then we abort as the violation has
+already been reported. Otherwise we create the account, with the slashing program
+as the owner. In this account we store the following:
+
+```rust
+struct ProofReport {
+  reporter: Pubkey,              // Fee payer, to allow the account to be closed
+  epoch: Epoch,                  // Epoch in which this report was created
+  pubkey: Pubkey,                // The pubkey of the node that committed the violation
+  slot: Slot,                    // Slot in which the violation occured
+  violation_type: u8,            // The violation type
+  proof: Vec<u8>                 // The serialized proof
+  proof_account: Option<Pubkey>, // Optional account where proof is stored instead
```

Comment:
Rather than using an `Option`, let's just always include this, with the caveat that the data at `proof_account` might get modified. And let's add it before `proof` to make it easier to find